File: src/main/java/nz/co/aetheric/sameness/reflection/MethodCallResult.java
package nz.co.aetheric.sameness.reflection;

import org.apache.commons.lang3.tuple.Pair;

import javax.validation.constraints.NotNull;

/**
 * This just holds the results of a pairwise MethodCall creation process.
 * <p>Author: <a href="http://gplus.to/tzrlk">Peter Cummuskey</a></p>
 */
class MethodCallResult<T> extends Pair<MethodCall<T>, MethodCall<T>> {

	private final MethodCall<T> left;
	private final MethodCall<T> right;

	MethodCallResult(@NotNull MethodCall<T> left, @NotNull MethodCall<T> right) {
		this.left = left;
		this.right = right;
	}

	@NotNull
	@Override
	public MethodCall<T> getLeft() {
		return left;
	}

	@NotNull
	@Override
	public MethodCall<T> getRight() {
		return right;
	}

	/** @deprecated */
	@Override
	@Deprecated
	public MethodCall<T> setValue(MethodCall<T> value) {
		throw new RuntimeException("Not allowed to edit immutable results.");
	}

}

File: src/main/kotlin/nz/co/aetheric/sameness/reflection/MethodCallResult.kt
package nz.co.aetheric.sameness.reflection

import org.apache.commons.lang3.tuple.Pair

import javax.validation.constraints.NotNull

/**
 * This just holds the results of a pairwise MethodCall creation process.
 *
 * Author: [Peter Cummuskey](http://gplus.to/tzrlk)
 */
internal class MethodCallResult<T> : Pair<MethodCall<T>, MethodCall<T>> {

	private val left: MethodCall<T>
	private val right: MethodCall<T>

	@Suppress("ConvertSecondaryConstructorToPrimary")
	constructor(left: MethodCall<T>, right: MethodCall<T>) : super() {
		this.left = left
		this.right = right
	}

	override fun getLeft()
			= left

	override fun getRight()
			= right

	override fun setValue(newValue: MethodCall<T>)
			= throw RuntimeException("Not allowed to edit immutable results.")

}
